In [1]:
import pandas as pd
import numpy as np

In [2]:
ted_data = pd.read_csv("ted_main.csv")
youtube_data = pd.read_csv("ted_youtube.csv")

In [3]:
ted_data.drop(['description','film_date','published_date','url'],axis=1,inplace=True)

In [4]:
ted_data=ted_data.rename({'comments':'ted_nof_comments','duration':'ted_duration','event':'ted_event','languages':'ted_avlb_lang','published_dt':
                'ted_published_dt','ratings':'ted_ratings','related_talks':'ted_talks_related','tags':'ted_tags','title':'ted_title','views':'ted_views'},axis=1)

In [5]:
youtube_data.drop(['Unnamed: 0','fulltitle','webpage_url','uploader_id','description','fulltitle_compare','id'],axis=1,inplace=True)

In [6]:
youtube_data = youtube_data.rename({'view_count':'youtube_views','duration':'youtube_duration','categories':'youtube_categories','tags':'youtube_tags',
                    'dislike_count':'youtube_dislikes','like_count':'youtube_likes','average_rating':'youtube_avg_ratings','title':'youtube_title'},axis=1)

In [7]:
speaker_occupation_na = ['Writer, educator, CEO','Entrepreneur, Politician','Technologist, CEO','Musician','Artist','Photographer, Artist, Activist']
ted_data.loc[ted_data.speaker_occupation.isna(),'speaker_occupation'] = speaker_occupation_na

In [8]:
youtube_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2492 entries, 0 to 2491
Data columns (total 8 columns):
youtube_views          2492 non-null int64
youtube_title          2492 non-null object
youtube_tags           2492 non-null object
youtube_categories     2492 non-null object
youtube_duration       2492 non-null int64
youtube_dislikes       2482 non-null float64
youtube_likes          2482 non-null float64
youtube_avg_ratings    2492 non-null float64
dtypes: float64(3), int64(2), object(3)
memory usage: 155.8+ KB


In [9]:
youtube_data['youtube_likes']=youtube_data['youtube_likes'].fillna(0)
youtube_data['youtube_dislikes']=youtube_data['youtube_dislikes'].fillna(0)

In [10]:
youtube_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2492 entries, 0 to 2491
Data columns (total 8 columns):
youtube_views          2492 non-null int64
youtube_title          2492 non-null object
youtube_tags           2492 non-null object
youtube_categories     2492 non-null object
youtube_duration       2492 non-null int64
youtube_dislikes       2492 non-null float64
youtube_likes          2492 non-null float64
youtube_avg_ratings    2492 non-null float64
dtypes: float64(3), int64(2), object(3)
memory usage: 155.8+ KB


In [11]:
youtube_data['melt1'] = youtube_data.youtube_title.str.split(':')

youtube_data['talk1']=youtube_data['melt1'].str.get(1)
youtube_data['author1']=youtube_data['melt1'].str.get(0)

In [12]:
youtube_data['melt2'] = youtube_data.youtube_title.str.split('|')

youtube_data['talk2']=youtube_data['melt2'].str.get(0)
youtube_data['author2']=youtube_data['melt2'].str.get(1)

In [13]:
youtube_data.head(5)

youtube_views                                      youtube_title  \
0          20695  Inge Missmahl brings peace to the minds of Afg...   
1         129968  Edith Widder: The weird and wonderful world of...   
2          46844     Jacqueline Novogratz on an escape from poverty   
3         281435            Math class needs a makeover | Dan Meyer   
4         192313              Craig Venter unveils "synthetic life"   

                                        youtube_tags  \
0  ['Inge', 'Missmahl', 'TEDTalks', 'TED', 'talks...   
1  ['Edith', 'Widder', 'TEDTalks', 'TED', 'talks'...   
2  ['Jacqueline', 'Novogratz', 'TEDTalks', 'TED',...   
3  ['Dan', 'Meyer', 'TEDTalks', 'TED', 'talks', '...   
4  ['TEDTalks', 'TED', 'talks', 'Craig Venter', '...   

          youtube_categories  youtube_duration  youtube_dislikes  \
0  ['Nonprofits & Activism']               830              14.0   
1   ['Science & Technology']               766              17.0   
2  ['Nonprofits & Activism']               451              42.0   
3   ['Science & Technology']               948              47.0   
4   ['Science & Technology']              1098             106.0   

   youtube_likes  youtube_avg_ratings  \
0          193.0             4.729469   
1         1155.0             4.941979   
2          325.0             4.542234   
3         3019.0             4.938683   
4         2468.0             4.835276   

                                               melt1  \
0  [Inge Missmahl brings peace to the minds of Af...   
1  [Edith Widder,  The weird and wonderful world ...   
2   [Jacqueline Novogratz on an escape from poverty]   
3          [Math class needs a makeover | Dan Meyer]   
4            [Craig Venter unveils "synthetic life"]   

                                               talk1  \
0                                                NaN   
1   The weird and wonderful world of bioluminescence   
2                                                NaN   
3                                                NaN   
4                                                NaN   

                                             author1  \
0  Inge Missmahl brings peace to the minds of Afg...   
1                                       Edith Widder   
2     Jacqueline Novogratz on an escape from poverty   
3            Math class needs a makeover | Dan Meyer   
4              Craig Venter unveils "synthetic life"   

                                               melt2  \
0  [Inge Missmahl brings peace to the minds of Af...   
1  [Edith Widder: The weird and wonderful world o...   
2   [Jacqueline Novogratz on an escape from poverty]   
3         [Math class needs a makeover ,  Dan Meyer]   
4            [Craig Venter unveils "synthetic life"]   

                                               talk2     author2  
0  Inge Missmahl brings peace to the minds of Afg...         NaN  
1  Edith Widder: The weird and wonderful world of...         NaN  
2     Jacqueline Novogratz on an escape from poverty         NaN  
3                       Math class needs a makeover    Dan Meyer  
4              Craig Venter unveils "synthetic life"         NaN

In [14]:
youtube_data['talk3']= youtube_data.youtube_title[(youtube_data.author2.isna()) & (youtube_data.talk1.isna())]

In [15]:
youtube_data['talk3']=youtube_data['talk3'].fillna(youtube_data['author1'].str.cat(youtube_data['talk1'],sep=' '))
youtube_data['talk3']=youtube_data['talk3'].fillna(youtube_data['author2'].str.cat(youtube_data['talk2'],sep=' '))

In [16]:
youtube_data.head(5)

youtube_views                                      youtube_title  \
0          20695  Inge Missmahl brings peace to the minds of Afg...   
1         129968  Edith Widder: The weird and wonderful world of...   
2          46844     Jacqueline Novogratz on an escape from poverty   
3         281435            Math class needs a makeover | Dan Meyer   
4         192313              Craig Venter unveils "synthetic life"   

                                        youtube_tags  \
0  ['Inge', 'Missmahl', 'TEDTalks', 'TED', 'talks...   
1  ['Edith', 'Widder', 'TEDTalks', 'TED', 'talks'...   
2  ['Jacqueline', 'Novogratz', 'TEDTalks', 'TED',...   
3  ['Dan', 'Meyer', 'TEDTalks', 'TED', 'talks', '...   
4  ['TEDTalks', 'TED', 'talks', 'Craig Venter', '...   

          youtube_categories  youtube_duration  youtube_dislikes  \
0  ['Nonprofits & Activism']               830              14.0   
1   ['Science & Technology']               766              17.0   
2  ['Nonprofits & Activism']               451              42.0   
3   ['Science & Technology']               948              47.0   
4   ['Science & Technology']              1098             106.0   

   youtube_likes  youtube_avg_ratings  \
0          193.0             4.729469   
1         1155.0             4.941979   
2          325.0             4.542234   
3         3019.0             4.938683   
4         2468.0             4.835276   

                                               melt1  \
0  [Inge Missmahl brings peace to the minds of Af...   
1  [Edith Widder,  The weird and wonderful world ...   
2   [Jacqueline Novogratz on an escape from poverty]   
3          [Math class needs a makeover | Dan Meyer]   
4            [Craig Venter unveils "synthetic life"]   

                                               talk1  \
0                                                NaN   
1   The weird and wonderful world of bioluminescence   
2                                                NaN   
3                                                NaN   
4                                                NaN   

                                             author1  \
0  Inge Missmahl brings peace to the minds of Afg...   
1                                       Edith Widder   
2     Jacqueline Novogratz on an escape from poverty   
3            Math class needs a makeover | Dan Meyer   
4              Craig Venter unveils "synthetic life"   

                                               melt2  \
0  [Inge Missmahl brings peace to the minds of Af...   
1  [Edith Widder: The weird and wonderful world o...   
2   [Jacqueline Novogratz on an escape from poverty]   
3         [Math class needs a makeover ,  Dan Meyer]   
4            [Craig Venter unveils "synthetic life"]   

                                               talk2     author2  \
0  Inge Missmahl brings peace to the minds of Afg...         NaN   
1  Edith Widder: The weird and wonderful world of...         NaN   
2     Jacqueline Novogratz on an escape from poverty         NaN   
3                       Math class needs a makeover    Dan Meyer   
4              Craig Venter unveils "synthetic life"         NaN   

                                               talk3  
0  Inge Missmahl brings peace to the minds of Afg...  
1  Edith Widder  The weird and wonderful world of...  
2     Jacqueline Novogratz on an escape from poverty  
3             Dan Meyer Math class needs a makeover   
4              Craig Venter unveils "synthetic life"

removing white space and changing all cases to loweer without spl char

In [17]:
youtube_data['talk3']=youtube_data['talk3'].str.replace('[^A-Za-z0-9]+','')
youtube_data['talk3']=youtube_data['talk3'].str.strip()
youtube_data['talk3']=youtube_data['talk3'].str.lower()

as ted name contains no : we are removing

In [18]:
ted_data['name']=ted_data['name'].str.replace(':','')
ted_data['name']=ted_data['name'].str.replace('[^A-Za-z0-9]+','')
ted_data['name']=ted_data['name'].str.strip()
ted_data['name']=ted_data['name'].str.lower()

in you tube data we have '(with english subtitles) in descr so removing

In [19]:
youtube_data.talk3[youtube_data['talk3'].str.contains('withenglish')]

536     carinamorillotounderstandautismdontlookawaywit...
661     karinagalperinshouldwesimplifyspellingwithengl...
687     jorgedrexlerpoetrymusicandidentitywithenglishs...
932     jorgeramoswhyjournalistshaveanobligationtochal...
996     emilianosalinasacivilresponsetoviolencewitheng...
1206    tomassaracenowouldyouliveinafloatingcityinthes...
1596    gabrielagonzlezhowligodiscoveredgravitationalw...
1859    rominalibsterthepowerofherdimmunitywithenglish...
2021    gonzalovilariohowargentinasblindsoccerteambeca...
2141    eduardosenzdecabeznmathisforeverwithenglishsub...
2351    cristinadomenechpoetrythatfreesthesoulwithengl...
Name: talk3, dtype: object

In [20]:
youtube_data['talk3']=youtube_data['talk3'].str.replace('withenglishsubtitles','')

In [21]:
pd.merge(ted_data,youtube_data,left_on='name',right_on='talk3').head(5)

ted_nof_comments  ted_duration ted_event    film_dt  ted_avlb_lang  \
0               265           977   TED2006  2/25/2006             43   
1               124          1286   TED2006  2/24/2006             26   
2               200          1116   TED2006  2/26/2006             35   
3               672          1305   TED2006   2/2/2006             36   
4               919           992   TED2006  2/24/2006             31   

    main_speaker                            name  num_speaker  \
0        Al Gore  algoreavertingtheclimatecrisis            1   
1    David Pogue       davidpoguesimplicitysells            1   
2  Majora Carter   majoracartergreeningtheghetto            1   
3   Tony Robbins      tonyrobbinswhywedowhatwedo            1   
4  Julia Sweeney      juliasweeneylettinggoofgod            1   

  ted_published_dt                                        ted_ratings  \
0        6/27/2006  [{'id': 7, 'name': 'Funny', 'count': 544}, {'i...   
1        6/27/2006  [{'id': 7, 'name': 'Funny', 'count': 964}, {'i...   
2        6/27/2006  [{'id': 3, 'name': 'Courageous', 'count': 760}...   
3        6/27/2006  [{'id': 7, 'name': 'Funny', 'count': 1102}, {'...   
4        7/10/2006  [{'id': 3, 'name': 'Courageous', 'count': 477}...   

                ...               youtube_dislikes youtube_likes  \
0               ...                          184.0         588.0   
1               ...                           46.0         549.0   
2               ...                           36.0         995.0   
3               ...                         3003.0       60423.0   
4               ...                          430.0        4343.0   

  youtube_avg_ratings                                    melt1  \
0            4.046632  [Averting the climate crisis | Al Gore]   
1            4.690756         [David Pogue,  Simplicity sells]   
2            4.860330    [Greening the ghetto | Majora Carter]   
3            4.810614    [Why we do what we do | Tony Robbins]   
4            4.639640      [Letting go of God | Julia Sweeney]   

               talk1                                author1  \
0                NaN  Averting the climate crisis | Al Gore   
1   Simplicity sells                            David Pogue   
2                NaN    Greening the ghetto | Majora Carter   
3                NaN    Why we do what we do | Tony Robbins   
4                NaN      Letting go of God | Julia Sweeney   

                                      melt2                          talk2  \
0  [Averting the climate crisis ,  Al Gore]   Averting the climate crisis    
1           [David Pogue: Simplicity sells]  David Pogue: Simplicity sells   
2    [Greening the ghetto ,  Majora Carter]           Greening the ghetto    
3    [Why we do what we do ,  Tony Robbins]          Why we do what we do    
4      [Letting go of God ,  Julia Sweeney]             Letting go of God    

          author2                           talk3  
0         Al Gore  algoreavertingtheclimatecrisis  
1             NaN       davidpoguesimplicitysells  
2   Majora Carter   majoracartergreeningtheghetto  
3    Tony Robbins      tonyrobbinswhywedowhatwedo  
4   Julia Sweeney      juliasweeneylettinggoofgod  

[5 rows x 30 columns]

soem descriptions differ by few workds so taking slice of text to match

In [22]:
ted_data['name_slice']= ted_data['name'].str.slice(start=0,stop=25)
youtube_data['talk_slice']=youtube_data['talk3'].str.slice(start=0,stop=25)

In [23]:
table_merge1 = pd.merge(ted_data,youtube_data,left_on='name_slice',right_on='talk_slice')
table_merge1.head(5)

ted_nof_comments  ted_duration ted_event    film_dt  ted_avlb_lang  \
0               265           977   TED2006  2/25/2006             43   
1               124          1286   TED2006  2/24/2006             26   
2               200          1116   TED2006  2/26/2006             35   
3               672          1305   TED2006   2/2/2006             36   
4               919           992   TED2006  2/24/2006             31   

    main_speaker                            name  num_speaker  \
0        Al Gore  algoreavertingtheclimatecrisis            1   
1    David Pogue       davidpoguesimplicitysells            1   
2  Majora Carter   majoracartergreeningtheghetto            1   
3   Tony Robbins      tonyrobbinswhywedowhatwedo            1   
4  Julia Sweeney      juliasweeneylettinggoofgod            1   

  ted_published_dt                                        ted_ratings  \
0        6/27/2006  [{'id': 7, 'name': 'Funny', 'count': 544}, {'i...   
1        6/27/2006  [{'id': 7, 'name': 'Funny', 'count': 964}, {'i...   
2        6/27/2006  [{'id': 3, 'name': 'Courageous', 'count': 760}...   
3        6/27/2006  [{'id': 7, 'name': 'Funny', 'count': 1102}, {'...   
4        7/10/2006  [{'id': 3, 'name': 'Courageous', 'count': 477}...   

             ...             youtube_likes youtube_avg_ratings  \
0            ...                     588.0            4.046632   
1            ...                     549.0            4.690756   
2            ...                     995.0            4.860330   
3            ...                   60423.0            4.810614   
4            ...                    4343.0            4.639640   

                                     melt1              talk1  \
0  [Averting the climate crisis | Al Gore]                NaN   
1         [David Pogue,  Simplicity sells]   Simplicity sells   
2    [Greening the ghetto | Majora Carter]                NaN   
3    [Why we do what we do | Tony Robbins]                NaN   
4      [Letting go of God | Julia Sweeney]                NaN   

                                 author1  \
0  Averting the climate crisis | Al Gore   
1                            David Pogue   
2    Greening the ghetto | Majora Carter   
3    Why we do what we do | Tony Robbins   
4      Letting go of God | Julia Sweeney   

                                      melt2                          talk2  \
0  [Averting the climate crisis ,  Al Gore]   Averting the climate crisis    
1           [David Pogue: Simplicity sells]  David Pogue: Simplicity sells   
2    [Greening the ghetto ,  Majora Carter]           Greening the ghetto    
3    [Why we do what we do ,  Tony Robbins]          Why we do what we do    
4      [Letting go of God ,  Julia Sweeney]             Letting go of God    

          author2                           talk3                 talk_slice  
0         Al Gore  algoreavertingtheclimatecrisis  algoreavertingtheclimatec  
1             NaN       davidpoguesimplicitysells  davidpoguesimplicitysells  
2   Majora Carter   majoracartergreeningtheghetto  majoracartergreeningthegh  
3    Tony Robbins      tonyrobbinswhywedowhatwedo  tonyrobbinswhywedowhatwed  
4   Julia Sweeney      juliasweeneylettinggoofgod  juliasweeneylettinggoofgo  

[5 rows x 32 columns]

to find the difference between the two dataframes we use outer join

In [24]:
join_outer = pd.merge(ted_data,youtube_data,left_on='name_slice',right_on='talk_slice',indicator=True,how='outer')

In [25]:
join_outer[join_outer['_merge']=='right_only'].head(5)

ted_nof_comments  ted_duration ted_event film_dt  ted_avlb_lang  \
2553               NaN           NaN       NaN     NaN            NaN   
2554               NaN           NaN       NaN     NaN            NaN   
2555               NaN           NaN       NaN     NaN            NaN   
2556               NaN           NaN       NaN     NaN            NaN   
2557               NaN           NaN       NaN     NaN            NaN   

     main_speaker name  num_speaker ted_published_dt ted_ratings     ...      \
2553          NaN  NaN          NaN              NaN         NaN     ...       
2554          NaN  NaN          NaN              NaN         NaN     ...       
2555          NaN  NaN          NaN              NaN         NaN     ...       
2556          NaN  NaN          NaN              NaN         NaN     ...       
2557          NaN  NaN          NaN              NaN         NaN     ...       

     youtube_avg_ratings                                              melt1  \
2553            4.729469  [Inge Missmahl brings peace to the minds of Af...   
2554            4.941979  [Edith Widder,  The weird and wonderful world ...   
2555            4.542234   [Jacqueline Novogratz on an escape from poverty]   
2556            4.835276            [Craig Venter unveils "synthetic life"]   
2557            4.811282  [Philip Howard,  Four ways to fix a broken leg...   

                                                  talk1  \
2553                                                NaN   
2554   The weird and wonderful world of bioluminescence   
2555                                                NaN   
2556                                                NaN   
2557             Four ways to fix a broken legal system   

                                                author1  \
2553  Inge Missmahl brings peace to the minds of Afg...   
2554                                       Edith Widder   
2555     Jacqueline Novogratz on an escape from poverty   
2556              Craig Venter unveils "synthetic life"   
2557                                      Philip Howard   

                                                  melt2  \
2553  [Inge Missmahl brings peace to the minds of Af...   
2554  [Edith Widder: The weird and wonderful world o...   
2555   [Jacqueline Novogratz on an escape from poverty]   
2556            [Craig Venter unveils "synthetic life"]   
2557  [Philip Howard: Four ways to fix a broken lega...   

                                                  talk2  author2  \
2553  Inge Missmahl brings peace to the minds of Afg...      NaN   
2554  Edith Widder: The weird and wonderful world of...      NaN   
2555     Jacqueline Novogratz on an escape from poverty      NaN   
2556              Craig Venter unveils "synthetic life"      NaN   
2557  Philip Howard: Four ways to fix a broken legal...      NaN   

                                                  talk3  \
2553     ingemissmahlbringspeacetothemindsofafghanistan   
2554  edithwiddertheweirdandwonderfulworldofbiolumin...   
2555           jacquelinenovogratzonanescapefrompoverty   
2556                    craigventerunveilssyntheticlife   
2557        philiphowardfourwaystofixabrokenlegalsystem   

                     talk_slice      _merge  
2553  ingemissmahlbringspeaceto  right_only  
2554  edithwiddertheweirdandwon  right_only  
2555  jacquelinenovogratzonanes  right_only  
2556  craigventerunveilssynthet  right_only  
2557  philiphowardfourwaystofix  right_only  

[5 rows x 33 columns]

some of the titles differ in ted and youtube data like "Eva Vertes - My dream about future medicine" is the titke in youtube but in ted 
it is "meet the future of camcer research"

We can try to match the remaining fields by author 

In [26]:
youtube_data['split']= youtube_data.talk2[(youtube_data['talk1'].isna())&(youtube_data['author2'].isna())].str.split(' ')

In [27]:
youtube_data['author3']=youtube_data['split'].str.get(0).str.cat(youtube_data['split'].str.get(1),sep=' ')

In [28]:
youtube_data.head(5)

youtube_views                                      youtube_title  \
0          20695  Inge Missmahl brings peace to the minds of Afg...   
1         129968  Edith Widder: The weird and wonderful world of...   
2          46844     Jacqueline Novogratz on an escape from poverty   
3         281435            Math class needs a makeover | Dan Meyer   
4         192313              Craig Venter unveils "synthetic life"   

                                        youtube_tags  \
0  ['Inge', 'Missmahl', 'TEDTalks', 'TED', 'talks...   
1  ['Edith', 'Widder', 'TEDTalks', 'TED', 'talks'...   
2  ['Jacqueline', 'Novogratz', 'TEDTalks', 'TED',...   
3  ['Dan', 'Meyer', 'TEDTalks', 'TED', 'talks', '...   
4  ['TEDTalks', 'TED', 'talks', 'Craig Venter', '...   

          youtube_categories  youtube_duration  youtube_dislikes  \
0  ['Nonprofits & Activism']               830              14.0   
1   ['Science & Technology']               766              17.0   
2  ['Nonprofits & Activism']               451              42.0   
3   ['Science & Technology']               948              47.0   
4   ['Science & Technology']              1098             106.0   

   youtube_likes  youtube_avg_ratings  \
0          193.0             4.729469   
1         1155.0             4.941979   
2          325.0             4.542234   
3         3019.0             4.938683   
4         2468.0             4.835276   

                                               melt1  \
0  [Inge Missmahl brings peace to the minds of Af...   
1  [Edith Widder,  The weird and wonderful world ...   
2   [Jacqueline Novogratz on an escape from poverty]   
3          [Math class needs a makeover | Dan Meyer]   
4            [Craig Venter unveils "synthetic life"]   

                                               talk1  \
0                                                NaN   
1   The weird and wonderful world of bioluminescence   
2                                                NaN   
3                                                NaN   
4                                                NaN   

                                             author1  \
0  Inge Missmahl brings peace to the minds of Afg...   
1                                       Edith Widder   
2     Jacqueline Novogratz on an escape from poverty   
3            Math class needs a makeover | Dan Meyer   
4              Craig Venter unveils "synthetic life"   

                                               melt2  \
0  [Inge Missmahl brings peace to the minds of Af...   
1  [Edith Widder: The weird and wonderful world o...   
2   [Jacqueline Novogratz on an escape from poverty]   
3         [Math class needs a makeover ,  Dan Meyer]   
4            [Craig Venter unveils "synthetic life"]   

                                               talk2     author2  \
0  Inge Missmahl brings peace to the minds of Afg...         NaN   
1  Edith Widder: The weird and wonderful world of...         NaN   
2     Jacqueline Novogratz on an escape from poverty         NaN   
3                       Math class needs a makeover    Dan Meyer   
4              Craig Venter unveils "synthetic life"         NaN   

                                               talk3  \
0     ingemissmahlbringspeacetothemindsofafghanistan   
1  edithwiddertheweirdandwonderfulworldofbiolumin...   
2           jacquelinenovogratzonanescapefrompoverty   
3                    danmeyermathclassneedsamakeover   
4                    craigventerunveilssyntheticlife   

                  talk_slice  \
0  ingemissmahlbringspeaceto   
1  edithwiddertheweirdandwon   
2  jacquelinenovogratzonanes   
3  danmeyermathclassneedsama   
4  craigventerunveilssynthet   

                                               split               author3  
0  [Inge, Missmahl, brings, peace, to, the, minds...         Inge Missmahl  
1                                                NaN                   NaN  
2  [Jacqueline, Novogratz, on, an, escape, from, ...  Jacqueline 

In [29]:
youtube_data['author3'] = youtube_data['author3'].fillna(youtube_data['author1'])
youtube_data['author3'] = youtube_data['author3'].fillna(youtube_data['author2'])

In [30]:
youtube_data.head(5)

youtube_views                                      youtube_title  \
0          20695  Inge Missmahl brings peace to the minds of Afg...   
1         129968  Edith Widder: The weird and wonderful world of...   
2          46844     Jacqueline Novogratz on an escape from poverty   
3         281435            Math class needs a makeover | Dan Meyer   
4         192313              Craig Venter unveils "synthetic life"   

                                        youtube_tags  \
0  ['Inge', 'Missmahl', 'TEDTalks', 'TED', 'talks...   
1  ['Edith', 'Widder', 'TEDTalks', 'TED', 'talks'...   
2  ['Jacqueline', 'Novogratz', 'TEDTalks', 'TED',...   
3  ['Dan', 'Meyer', 'TEDTalks', 'TED', 'talks', '...   
4  ['TEDTalks', 'TED', 'talks', 'Craig Venter', '...   

          youtube_categories  youtube_duration  youtube_dislikes  \
0  ['Nonprofits & Activism']               830              14.0   
1   ['Science & Technology']               766              17.0   
2  ['Nonprofits & Activism']               451              42.0   
3   ['Science & Technology']               948              47.0   
4   ['Science & Technology']              1098             106.0   

   youtube_likes  youtube_avg_ratings  \
0          193.0             4.729469   
1         1155.0             4.941979   
2          325.0             4.542234   
3         3019.0             4.938683   
4         2468.0             4.835276   

                                               melt1  \
0  [Inge Missmahl brings peace to the minds of Af...   
1  [Edith Widder,  The weird and wonderful world ...   
2   [Jacqueline Novogratz on an escape from poverty]   
3          [Math class needs a makeover | Dan Meyer]   
4            [Craig Venter unveils "synthetic life"]   

                                               talk1  \
0                                                NaN   
1   The weird and wonderful world of bioluminescence   
2                                                NaN   
3                                                NaN   
4                                                NaN   

                                             author1  \
0  Inge Missmahl brings peace to the minds of Afg...   
1                                       Edith Widder   
2     Jacqueline Novogratz on an escape from poverty   
3            Math class needs a makeover | Dan Meyer   
4              Craig Venter unveils "synthetic life"   

                                               melt2  \
0  [Inge Missmahl brings peace to the minds of Af...   
1  [Edith Widder: The weird and wonderful world o...   
2   [Jacqueline Novogratz on an escape from poverty]   
3         [Math class needs a makeover ,  Dan Meyer]   
4            [Craig Venter unveils "synthetic life"]   

                                               talk2     author2  \
0  Inge Missmahl brings peace to the minds of Afg...         NaN   
1  Edith Widder: The weird and wonderful world of...         NaN   
2     Jacqueline Novogratz on an escape from poverty         NaN   
3                       Math class needs a makeover    Dan Meyer   
4              Craig Venter unveils "synthetic life"         NaN   

                                               talk3  \
0     ingemissmahlbringspeacetothemindsofafghanistan   
1  edithwiddertheweirdandwonderfulworldofbiolumin...   
2           jacquelinenovogratzonanescapefrompoverty   
3                    danmeyermathclassneedsamakeover   
4                    craigventerunveilssyntheticlife   

                  talk_slice  \
0  ingemissmahlbringspeaceto   
1  edithwiddertheweirdandwon   
2  jacquelinenovogratzonanes   
3  danmeyermathclassneedsama   
4  craigventerunveilssynthet   

                                               split  \
0  [Inge, Missmahl, brings, peace, to, the, minds...   
1                                                NaN   
2  [Jacqueline, Novogratz, on, an, escape, from, ...   
3                                                NaN   
4        [Craig,

Now by separating the unique column in ted data and youtube data then we can join by authors

In [31]:
 pd.merge(ted_data,youtube_data,left_on='name_slice',right_on='talk_slice').head(5)

ted_nof_comments  ted_duration ted_event    film_dt  ted_avlb_lang  \
0               265           977   TED2006  2/25/2006             43   
1               124          1286   TED2006  2/24/2006             26   
2               200          1116   TED2006  2/26/2006             35   
3               672          1305   TED2006   2/2/2006             36   
4               919           992   TED2006  2/24/2006             31   

    main_speaker                            name  num_speaker  \
0        Al Gore  algoreavertingtheclimatecrisis            1   
1    David Pogue       davidpoguesimplicitysells            1   
2  Majora Carter   majoracartergreeningtheghetto            1   
3   Tony Robbins      tonyrobbinswhywedowhatwedo            1   
4  Julia Sweeney      juliasweeneylettinggoofgod            1   

  ted_published_dt                                        ted_ratings  \
0        6/27/2006  [{'id': 7, 'name': 'Funny', 'count': 544}, {'i...   
1        6/27/2006  [{'id': 7, 'name': 'Funny', 'count': 964}, {'i...   
2        6/27/2006  [{'id': 3, 'name': 'Courageous', 'count': 760}...   
3        6/27/2006  [{'id': 7, 'name': 'Funny', 'count': 1102}, {'...   
4        7/10/2006  [{'id': 3, 'name': 'Courageous', 'count': 477}...   

                   ...                    \
0                  ...                     
1                  ...                     
2                  ...                     
3                  ...                     
4                  ...                     

                                     melt1              talk1  \
0  [Averting the climate crisis | Al Gore]                NaN   
1         [David Pogue,  Simplicity sells]   Simplicity sells   
2    [Greening the ghetto | Majora Carter]                NaN   
3    [Why we do what we do | Tony Robbins]                NaN   
4      [Letting go of God | Julia Sweeney]                NaN   

                                 author1  \
0  Averting the climate crisis | Al Gore   
1                            David Pogue   
2    Greening the ghetto | Majora Carter   
3    Why we do what we do | Tony Robbins   
4      Letting go of God | Julia Sweeney   

                                      melt2                          talk2  \
0  [Averting the climate crisis ,  Al Gore]   Averting the climate crisis    
1           [David Pogue: Simplicity sells]  David Pogue: Simplicity sells   
2    [Greening the ghetto ,  Majora Carter]           Greening the ghetto    
3    [Why we do what we do ,  Tony Robbins]          Why we do what we do    
4      [Letting go of God ,  Julia Sweeney]             Letting go of God    

          author2                           talk3                 talk_slice  \
0         Al Gore  algoreavertingtheclimatecrisis  algoreavertingtheclimatec   
1             NaN       davidpoguesimplicitysells  davidpoguesimplicitysells   
2   Majora Carter   majoracartergreeningtheghetto  majoracartergreeningthegh   
3    Tony Robbins      tonyrobbinswhywedowhatwedo  tonyrobbinswhywedowhatwed   
4   Julia Sweeney      juliasweeneylettinggoofgod  juliasweeneylettinggoofgo   

  split                                author3  
0   NaN  Averting the climate crisis | Al Gore  
1   NaN                            David Pogue  
2   NaN    Greening the ghetto | Majora Carter  
3   NaN    Why we do what we do | Tony Robbins  
4   NaN      Letting go of God | Julia Sweeney  

[5 rows x 34 columns]

Youtube data have 2492 rows but the mapped no of columns are 1993. So using outer join to find unmatched rows

In [32]:
data_join_outer = pd.merge(ted_data,youtube_data,left_on='name_slice',right_on='talk_slice',indicator=True,how='outer')

In [33]:
uni_youtube = data_join_outer[data_join_outer['_merge']=='right_only']

In [34]:
uni_youtube.columns

Index(['ted_nof_comments', 'ted_duration', 'ted_event', 'film_dt',
       'ted_avlb_lang', 'main_speaker', 'name', 'num_speaker',
       'ted_published_dt', 'ted_ratings', 'ted_talks_related',
       'speaker_occupation', 'ted_tags', 'ted_title', 'ted_views',
       'name_slice', 'youtube_views', 'youtube_title', 'youtube_tags',
       'youtube_categories', 'youtube_duration', 'youtube_dislikes',
       'youtube_likes', 'youtube_avg_ratings', 'melt1', 'talk1', 'author1',
       'melt2', 'talk2', 'author2', 'talk3', 'talk_slice', 'split', 'author3',
       '_merge'],
      dtype='object')

In [35]:
uni_youtube.drop(['ted_nof_comments', 'ted_duration', 'ted_event',
       'film_dt', 'ted_avlb_lang', 'main_speaker', 'name', 'num_speaker',
        'ted_published_dt', 'ted_ratings', 'ted_talks_related',
       'speaker_occupation', 'ted_tags', 'ted_title', 'name_slice'],inplace=True,axis=1)

C:\Users\Sivakumar\Anaconda3\lib\site-packages\ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  after removing the cwd from sys.path.


In [36]:
uni_ted = data_join_outer[data_join_outer['_merge']=='left_only']

In [37]:
uni_ted.columns

Index(['ted_nof_comments', 'ted_duration', 'ted_event', 'film_dt',
       'ted_avlb_lang', 'main_speaker', 'name', 'num_speaker',
       'ted_published_dt', 'ted_ratings', 'ted_talks_related',
       'speaker_occupation', 'ted_tags', 'ted_title', 'ted_views',
       'name_slice', 'youtube_views', 'youtube_title', 'youtube_tags',
       'youtube_categories', 'youtube_duration', 'youtube_dislikes',
       'youtube_likes', 'youtube_avg_ratings', 'melt1', 'talk1', 'author1',
       'melt2', 'talk2', 'author2', 'talk3', 'talk_slice', 'split', 'author3',
       '_merge'],
      dtype='object')

In [38]:
uni_ted.drop(['youtube_views', 'youtube_title', 'youtube_categories', 'youtube_duration',
       'youtube_dislikes', 'youtube_likes', 'youtube_avg_ratings',
       'melt1', 'talk1', 'author1', 'melt2', 'author2', 'talk2', 'talk3',
       'talk_slice', '_merge'],inplace=True,axis=1)

C:\Users\Sivakumar\Anaconda3\lib\site-packages\ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  after removing the cwd from sys.path.


In [39]:
data_merge = pd.merge(uni_ted,uni_youtube,left_on='main_speaker',right_on='author3')

In [40]:
data_merge.head(5)

ted_nof_comments  ted_duration ted_event    film_dt  ted_avlb_lang  \
0             593.0        1190.0   TED2006  2/22/2006           48.0   
1             125.0         602.0   TED2009   2/6/2009           40.0   
2              46.0        1198.0   TED2006  2/23/2006           19.0   
3             852.0        1485.0   TED2006   2/2/2006           32.0   
4              79.0        1414.0   TED2006  2/26/2006           27.0   

          main_speaker                                               name  \
0         Hans Rosling               hansroslingthebeststatsyouveeverseen   
1         Hans Rosling      hansroslinginsightsonhivinstunningdatavisuals   
2  Joshua Prince-Ramus  joshuaprinceramusbehindthedesignofseattleslibrary   
3          Dan Dennett    dandennettletsteachreligionallreligioninschools   
4     Cameron Sinclair  cameronsinclairmywishacallforopensourcearchite...   

   num_speaker ted_published_dt  \
0          1.0        6/27/2006   
1          1.0        5/13/2009   
2          1.0        7/10/2006   
3          1.0        7/18/2006   
4          1.0        7/25/2006   

                                         ted_ratings     ...      \
0  [{'id': 9, 'name': 'Ingenious', 'count': 3202}...     ...       
1  [{'id': 23, 'name': 'Jaw-dropping', 'count': 9...     ...       
2  [{'id': 9, 'name': 'Ingenious', 'count': 195},...     ...       
3  [{'id': 3, 'name': 'Courageous', 'count': 562}...     ...       
4  [{'id': 3, 'name': 'Courageous', 'count': 169}...     ...       

                                               talk1              author1  \
0   Debunking third-world myths with the best sta...         Hans Rosling   
1   Debunking third-world myths with the best sta...         Hans Rosling   
2              Designing the Seattle Central Library  Joshua Prince-Ramus   
3                   Responding to Pastor Rick Warren          Dan Dennett   
4                A call for open-source architecture     Cameron Sinclair   

                                               melt2  \
0  [Hans Rosling: Debunking third-world myths wit...   
1  [Hans Rosling: Debunking third-world myths wit...   
2  [Joshua Prince-Ramus: Designing the Seattle Ce...   
3    [Dan Dennett: Responding to Pastor Rick Warren]   
4  [Cameron Sinclair: A call for open-source arch...   

                                               talk2  author2  \
0  Hans Rosling: Debunking third-world myths with...      NaN   
1  Hans Rosling: Debunking third-world myths with...      NaN   
2  Joshua Prince-Ramus: Designing the Seattle Cen...      NaN   
3      Dan Dennett: Responding to Pastor Rick Warren      NaN   
4  Cameron Sinclair: A call for open-source archi...      NaN   

                                               talk3  \
0  hansroslingdebunkingthirdworldmythswiththebest...   
1  hansroslingdebunkingthirdworldmythswiththebest...   
2  joshuaprinceramusdesigningtheseattlecentrallib...   
3             dandennettrespondingtopastorrickwarren   
4      cameronsinclairacallforopensourcearchitecture   

                  talk_slice split_y            author3_y      _merge  
0  hansroslingdebunkingthird     NaN         Hans Rosling  right_only  
1  hansroslingdebunkingthird     NaN         Hans Rosling  right_only  
2  joshuaprinceramusdesignin     NaN  Joshua Prince-Ramus  right_only  
3  dandennettrespondingtopas     NaN          Dan Dennett  right_only  
4  cameronsinclairacallforop     NaN     Cameron Sinclair  right_only  

[5 rows x 39 columns]

In [41]:
data_merge[['talk_slice','name_slice']].head(5)

talk_slice                 name_slice
0  hansroslingdebunkingthird  hansroslingthebeststatsyo
1  hansroslingdebunkingthird  hansroslinginsightsonhivi
2  joshuaprinceramusdesignin  joshuaprinceramusbehindth
3  dandennettrespondingtopas  dandennettletsteachreligi
4  cameronsinclairacallforop  cameronsinclairmywishacal

While split and merging there may be chances of more than one talk for a speaker, so filtering of one talk by one speaker

In [42]:
data_merge.groupby('author3_y')['name'].count().sort_values(ascending=False).head(5)

author3_y
Steven Johnson      4
Jay Walker          4
Murray Gell-Mann    4
Craig Venter        4
Golan Levin         4
Name: name, dtype: int64

In [43]:
single_talk = data_merge.groupby('author3_y').filter(lambda g:g['name'].count()==1)

In [44]:
single_talk.head(5)

ted_nof_comments  ted_duration ted_event    film_dt  ted_avlb_lang  \
2               46.0        1198.0   TED2006  2/23/2006           19.0   
3              852.0        1485.0   TED2006   2/2/2006           32.0   
4               79.0        1414.0   TED2006  2/26/2006           27.0   
5               55.0        1538.0   TED2006  2/26/2006           20.0   
10              88.0         906.0   TED2006  2/24/2006           27.0   

           main_speaker                                               name  \
2   Joshua Prince-Ramus  joshuaprinceramusbehindthedesignofseattleslibrary   
3           Dan Dennett    dandennettletsteachreligionallreligioninschools   
4      Cameron Sinclair  cameronsinclairmywishacallforopensourcearchite...   
5        Jehane Noujaim                jehanenoujaimmywishaglobaldayoffilm   
10            Amy Smith                   amysmithsimpledesignstosavealife   

    num_speaker ted_published_dt  \
2           1.0        7/10/2006   
3           1.0        7/18/2006   
4           1.0        7/25/2006   
5           1.0        7/25/2006   
10          1.0        8/15/2006   

                                          ted_ratings     ...      \
2   [{'id': 9, 'name': 'Ingenious', 'count': 195},...     ...       
3   [{'id': 3, 'name': 'Courageous', 'count': 562}...     ...       
4   [{'id': 3, 'name': 'Courageous', 'count': 169}...     ...       
5   [{'id': 1, 'name': 'Beautiful', 'count': 70}, ...     ...       
10  [{'id': 9, 'name': 'Ingenious', 'count': 321},...     ...       

                                                talk1              author1  \
2               Designing the Seattle Central Library  Joshua Prince-Ramus   
3                    Responding to Pastor Rick Warren          Dan Dennett   
4                 A call for open-source architecture     Cameron Sinclair   
5                                       TEDPrize wish       Jehane Noujaim   
10   Simple designs that could save millions of ch...            Amy Smith   

                                                melt2  \
2   [Joshua Prince-Ramus: Designing the Seattle Ce...   
3     [Dan Dennett: Responding to Pastor Rick Warren]   
4   [Cameron Sinclair: A call for open-source arch...   
5   [Jehane Noujaim: TEDPrize wish: Unite the worl...   
10  [Amy Smith: Simple designs that could save mil...   

                                                talk2  author2  \
2   Joshua Prince-Ramus: Designing the Seattle Cen...      NaN   
3       Dan Dennett: Responding to Pastor Rick Warren      NaN   
4   Cameron Sinclair: A call for open-source archi...      NaN   
5   Jehane Noujaim: TEDPrize wish: Unite the world...      NaN   
10  Amy Smith: Simple designs that could save mill...      NaN   

                                                talk3  \
2   joshuaprinceramusdesigningtheseattlecentrallib...   
3              dandennettrespondingtopastorrickwarren   
4       cameronsinclairacallforopensourcearchitecture   
5                           jehanenoujaimtedprizewish   
10  amysmithsimpledesignsthatcouldsavemillionsofch...   

                   talk_slice split_y            author3_y      _merge  
2   joshuaprinceramusdesignin     NaN  Joshua Prince-Ramus  right_only  
3   dandennettrespondingtopas     NaN          Dan Dennett  right_only  
4   cameronsinclairacallforop     NaN     Cameron Sinclair  right_only  
5   jehanenoujaimtedprizewish     NaN       Jehane Noujaim  right_only  
10  amysmithsimpledesignsthat     NaN            Amy Smith  right_only  

[5 rows x 39 columns]

    So now 1993 + 259 we can able to match upto 2252 rows

In [45]:
youtube_data[['youtube_title','youtube_views']].sort_values(by='youtube_views',ascending = False).head(10)

youtube_title  youtube_views
1900  This is what happens when you reply to spam em...       17941227
971   How to speak so that people want to listen | J...       14398424
2468     Do schools kill creativity? | Sir Ken Robinson       13838924
942   Your body language may shape who you are | Amy...       13402518
842   Looks aren't everything. Believe me, I'm a mod...       13139379
81                  Why we do what we do | Tony Robbins       11424257
1930  How I held my breath for 17 minutes | David Bl...       11160827
2195           The art of misdirection | Apollo Robbins       10531729
414   Inside the mind of a master procrastinator | T...        9863319
821                   How to spot a liar | Pamela Meyer        9785607

In [ ]:
youtube_data.head(5)

In [ ]:
type(youtube_data.youtube_categories[0])

In [ ]:
ted_data.head(5)

In [47]:

import json

In [48]:
import ast
df_list=[]

for index,row in ted_data.iterrows():
    ratings_list = ast.literal_eval(row.ted_ratings)
    ratings_df = pd.DataFrame(ratings_list)
    ratings_df['ted_title']=row.ted_title
    df_list.append(ratings_df)
    
ratings = pd.concat(df_list,axis=0)

In [49]:
ratings.head(5)

count  id        name                    ted_title
0  19645   7       Funny  Do schools kill creativity?
1   4573   1   Beautiful  Do schools kill creativity?
2   6073   9   Ingenious  Do schools kill creativity?
3   3253   3  Courageous  Do schools kill creativity?
4    387  11  Longwinded  Do schools kill creativity?